proctor-common/src/main/java/com/indeed/proctor/common/ProctorUtils.java

```diff
@@ -1146,6 +1159,19 @@ private static void verifyInternallyConsistentDefinition(
                 throw new IncompatibleTestMatrixException(
                         "Allocation range has no buckets, needs to add up to 1.");
             }
+            if (testDefinition.getEnableUnitlessAllocations() && allocation.getRule() != null) {
+                final int unitlessIdentifierIndex =
+                        allocation.getRule().indexOf(UNITLESS_ALLOCATION_IDENTIFIER);
+
+                if (unitlessIdentifierIndex != -1
+                        && (unitlessIdentifierIndex == 0
+                                || allocation.getRule().charAt(unitlessIdentifierIndex - 1) != '!')
```

Comment:
can someone write a rule as missingExperimentalUnit == false?

Reply:
can the logic inside containsUnitlessAllocation be reused here? or vice versa

Reply:
Add == false check and made functions more reusable.

Reply:
overall i am nervous about parsing rule as it can be very complicated and has many variations.

missingExperimentalUnit
! missingExperimentalUnit
missingExperimentalUnit == true
missingExperimentalUnit != true
missingExperimentalUnit == false
missingExperimentalUnit != false
!(missingExperimentalUnit == true)
!(missingExperimentalUnit == false)
...etc

I have a quick question, is missingExperimentalUnit a property that users can use to build a rule on Proctor UI? Or can it only be added by the Proctor UI programmatically when checking some box?

Reply:
I originally suggested using a FE checkbox to make allocations able to be unitless and not use rules at all but this was preferred by Brian as he wants less FE changes. 

Since the validation of missingExperimentalUnit is only done at artifact load and webapp CRUD validation, we could turn this into a regex. However the cases as you describe are somewhat endless. 

Side topic: Ideally our rule builder or rule validation would also perform Boolean Algebra laws to simplify rules before saving them. This would save time on evaluation and memory space as rules would be shortened. Like for example there is no reason to save a rule with booleanValue == true. That is 8 characters longer than it needs to be.

Reply:
As per convo with Brian in stand up on 8/31 can just support basic boolean cases:

- missingExperimentalUnit
- !missingExperimentalUnit
- missingExperimentalUnit == true
- missingExperimentalUnit == false. 

Logic is:

```
    private static boolean isUnitlessAllocation(@Nonnull final Allocation allocation) {
        final int unitlessIdentifierIndex =
                allocation.getRule().indexOf(UNITLESS_ALLOCATION_IDENTIFIER);
        final boolean unitlessButFalse =
                allocation.getRule().contains(UNITLESS_ALLOCATION_IDENTIFIER + " == false");
        return !unitlessButFalse
                && unitlessIdentifierIndex != -1
                && (unitlessIdentifierIndex == 0
                        || allocation.getRule().charAt(unitlessIdentifierIndex - 1) != '!');
    }
```